### PR TITLE
[CALCITE-6478] JSON functions should return NULL when input is NULL

### PR DESCRIPTION
**Pull request by NobiGo**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 1]
```diff
@@ -3682,9 +3682,22 @@ static AbstractRexCallImplementor of(AbstractRexCallImplementor implementor) {
 
     @Override Expression implementSafe(final RexToLixTranslator translator,
         final RexCall call, final List<Expression> argValueList) {
+      // E.g., "final Boolean resultValue = (callValue == null) ? null : !callValue"
       final Expression expression =
           implementor.implementSafe(translator, call, argValueList);
-      return Expressions.not(expression);
+      final ParameterExpression callValue =
```
[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3688]: Hi @mihaibudiu, this is related to the `IS NOT JSON VALUE` function.  In Calcite, implement this function  by:

```
map.put(IS_NOT_JSON_VALUE,
          NotImplementor.of(
              new MethodImplementor(BuiltInMethod.IS_JSON_VALUE.method,
                  NullPolicy.NONE, false)));
```
NoT is implemented through NotImplementer.
Before this PR, the JsonFunctions‘ isJsonValue function is always TRUE or FALSE. So the NotImplementor only needs to do `!(isJsonValue )`. 
But After this PR, JsonFunctions‘ isJsonValue will return NULL. So the NotImplementor needs to change to `isJsonValue == null? isJsonValue : !(isJsonValue ) ` make sure `null IS NOT JSON VALUE` returns null.

[the remark this answers is not in the document]
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3688]: The isn't the bug in the contents of the map, and not in the NotImplementor? There should be a call in between which converts null to false. I imagine this pattern must show up in other places.

[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3688]: Here is the reason. 
```
  public static boolean isJsonValue(String input) {
    try {
      dejsonize(input);
      return true;
    } catch (Exception e) {
      return false;
    }
  }
```
If the input is NULL, dejsonize will throw an exception and return false.

[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3688]: That's why I added this implementation in isJsonValue.
```
if (input == null) {
      return null;
    }
```
to make isJsonValue return NULL when the input is NULL.

[the remark this answers is not in the document]
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3688]: If NotImplementor is not buggy we should not change it

[the remark this answers is not in the document]
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3688]: Isn't the bug here?

 NotImplementor.of(
              new MethodImplementor(BuiltInMethod.IS_JSON_VALUE.method,
                  NullPolicy.NONE, false)))

[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3688]: Yes, the NotImplementor can't handle when BuiltInMethod.IS_JSON_VALUE.method return null. Before this PR. If BuiltInMethod.IS_JSON_VALUE.method return null then NotImplementor  will generate `!NULL`, this is invalid. After this PR, NotImplementor will generate `jsonValue==null? null : !jsonValue`

[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3688]: @mihaibudiu Hi, Sorry to bother you, do you have any other questions or suggestions about this?

[the remark this answers is not in the document]
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3688]: NotImplementor is used in many other places, and you are changing the semantics of all of these.
Wouldn't it be simpler to just fix the expression computing IS_NOT_JSON_VALUE?

[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3688]: Ok. I understand your worries. I will use another NotJsonImplementor for this.

**Review by mihaibudiu — APPROVED** [review]

**Comment by mihaibudiu** [comment]
I think you can squash the commits.

**Pull request merged**